docs/design/auditing-containment.md

```diff
@@ -0,0 +1,168 @@
+```
+downloader, err := ps.Download(timedCtx, limit.GetLimit(), offset, int64(shareSize))
+```
+or
+```
+_, err = io.ReadFull(downloader, buf)
+```
+Once these errors have occurred and we know that a `pendingAudit` entry should be made in the ContainmentDB, we should use infectious’s `f.Decode` to generate the original stripe.
+Then we should use infectious’s `f.EncodeSingle` where we input the stripe and output the missing share. We then make a hash of that missing share to be saved to the ContainmentDB as `ExpectedShareHash`.
+
+`containment.Contain` should be called when a node is found to be “uncooperative."
+
+``` pkg/audit/verifier.go
+verifier.containment.Contain(ctx, &pendingAudit{
+    nodeID:            nodeID,
+    pieceID:           pieceID,
+    pieceNum:          pieceNum,
+    stripeIndex:       stripeIndex,
+    shareSize:         shareSize,
+    expectedShareHash: shareHash,
+}
+```
+
+#### 2. Audits continue normally. But now the nodes listed in a pointer are checked for `contained` status.
+``` pkg/audit/containment/checker.go
+func (verifier *Verifier) Verify(ctx context.Context, stripe *Stripe) (verifiedNodes *RecordAuditsInfo, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	pointer := stripe.Segment
+
+	pieces := pointer.GetRemote().GetRemotePieces()
+	
+	// get the NodeDossiers from the overlay using the pieces’ nodeIDs
+  // determine if they have the contained flag set to true
+  ...
+```
+
+#### 3. Contained nodes are checked for the same data that they were originally requested to respond with:
+
+```
+func (verifier *Verifier) Reverify(ctx context.Context, node storj.NodeID) (auditRecords *RecordAuditsInfo, err error) {
+
+  pendingAudit, err := checker.containment.Get(id)
+
+  // dial the node, then use info from the pendingAudit to download the target share from the node
+
+  offset := node.stripeIndex * node.shareSize
+  downloader, err := ps.Download(timedCtx, limit.GetLimit(), offset, int64(shareSize))
+
+  // if the error is a timeout (or any other error?), call ReverifyFail
+  // if the download occurred successfully, then get the hash of the original erasure share from the ContainmentDB
+
+  shareData := make([]byte, shareSize)
+  _, err = io.ReadFull(downloader, shareData)
+
+  // create a hash of shareData
+
+  var successNodes storj.NodeIDList
+  var failNodes storj.NodeIDList
+
+  if !bytes.Equal(shareHash, pendingAudit.expectedShareHash) {
+    failNodes = append(failNodes, node)
+    auditRecords.FailNodeIDs = failNodes
+  } else {
+    successNodes = append(successNodes, node)
+    auditRecords.SuccessNodeIDs = successNodes
+  }
+
+  // remove the set the contained flag to false on the NodeDossier and update the overlay (satellitedb)
+  // remove the pending audit from the ContainmentDB
+
+  return auditRecords
+```
+
+## Open Issues
+
+#### Why not audit a contained node for other data?
```

Comment:
Can we do both?  Rather than storing a list of failed audits, maybe if they fail the subsequent audits we just increment the failure count even though its different data?

Reply:
After our conversation this morning, it seemed like it was agreed to be a good idea that a contained node could be audited for other data as well. My initial thought was that we should keep track of which reverification failures correlate to which data to reduce possible confusion/misdirection about which pieces are missing, but just incrementing reverification failures could also be a quicker route to finding if a node needs to be disqualified. Would be open to more thoughts here.